File: container/hooks/lint-cpp-diff.py
#!/usr/bin/env python3
"""
PostToolUse lint hook for Edit/Write on C++ files.

Checks the new content (from Edit's new_string or Write's content) against
mechanical rules that are always wrong in Unreal Engine C++. Returns feedback
lines that the agent sees immediately.

Input: JSON on stdin with tool_input containing either:
  - new_string (Edit tool)
  - content (Write tool)
  - file_path

Exit 0: no issues (tool proceeds)
Exit 0 with stdout: issues found (tool proceeds, agent sees feedback)
We never block the write — just surface the problems immediately.
"""

import json
import re
import sys


def check_lines(lines: list[str], file_path: str) -> list[str]:
    issues: list[str] = []

    # Track multiline state for IILE detection
    full_text = "\n".join(lines)

    for i, line in enumerate(lines, start=1):
        stripped = line.strip()

        # Skip comments and preprocessor
        if stripped.startswith("//") or stripped.startswith("#") or stripped.startswith("/*") or stripped.startswith("*"):
            continue

        # Rule 1: East-const violation
        # Match "const Type&" or "const Type*" but not "const_cast" or "constexpr"
        if re.search(r"\bconst\s+(?!cast|expr|eval)\w+[\s*&]", line):
            # Exclude "const char*" in string literal contexts and return type positions
            # that are const-pointer-to-const (e.g. "FSlateBrush const* const")
            match = re.search(r"\bconst\s+(\w+)\s*[&*]", line)
            if match:
                typename = match.group(1)
                # Skip if it's a const pointer itself (const int* const)
                if typename not in ("cast", "expr", "eval", "override", "noexcept"):
                    issues.append(
                        f"  LINT [{file_path}:{i}] East-const: "
                        f"'const {typename}' should be '{typename} const'. "
                        f"Line: {stripped[:80]}"
                    )

        # Rule 2: Greedy lambda captures [&] or [=]
        if re.search(r"\[&\]\s*\(", line) or re.search(r"\[=\]\s*\(", line):
            issues.append(
                f"  LINT [{file_path}:{i}] Greedy capture: "
                f"use explicit captures instead of [&] or [=]. "
                f"Line: {stripped[:80]}"
            )
        # Also catch [&] and [=] without parens (no-arg lambdas)
        elif re.search(r"\[&\]\s*\{", line) or re.search(r"\[=\]\s*\{", line):
            issues.append(
                f"  LINT [{file_path}:{i}] Greedy capture: "
                f"use explicit captures instead of [&] or [=]. "
                f"Line: {stripped[:80]}"
            )

        # Rule 3: Raw new (outside blessed functions)
        # Match "new TypeName" but exclude NewObject, MakeShared, MakeUnique,
        # CreateDefaultSubobject, placement new, and operator new
        # Strip string literals before checking for raw new
        line_no_strings = re.sub(r'TEXT\s*\([^)]*\)', '', re.sub(r'"[^"]*"', '', line))
        new_match = re.search(r"\bnew\s+([A-Z]\w+)", line_no_strings)
        if new_match:
            # Check the broader context — is this inside a blessed call?
            blessed = ("NewObject", "MakeShared", "MakeUnique", "MakeShareable",
                       "CreateDefaultSubobject", "placement", "operator")
            if not any(b in line_no_strings for b in blessed):
                issues.append(
                    f"  LINT [{file_path}:{i}] Raw new: "
                    f"use NewObject<T>, MakeShared<T>, or MakeUnique<T> instead. "
                    f"Line: {stripped[:80]}"
                )

        # Rule 4: Multiple declarations on one line
        # Detect "Type Name1, Name2;" pattern (simplified)
        if re.search(r"^\s*\w[\w:<>*&\s]+\s+\w+\s*,\s*\w+\s*[;=]", line):
            # Exclude function parameter lists, for-loop inits, and template parameter lists
            if ("(" not in line.split(",")[0]
                    and not stripped.startswith("for")
                    and not stripped.startswith("template")):
                issues.append(
                    f"  LINT [{file_path}:{i}] Multiple declarations: "
                    f"declare one symbol per line. "
                    f"Line: {stripped[:80]}"
                )

        # Rule 5: Uninitialised TSharedRef member field
        # TSharedRef<SomeType> FieldName; (no = or { initialiser)
        if re.search(r"\bTSharedRef\s*<[^>]+>\s+\w+\s*;", line):
            if "=" not in line and "{" not in line:
                issues.append(
                    f"  LINT [{file_path}:{i}] Uninitialised TSharedRef: "
                    f"TSharedRef has no null state — initialise with MakeShared<T>() "
                    f"or use TSharedPtr if initialisation is deferred. "
                    f"Line: {stripped[:80]}"
                )

    # Rule 6: IILE detection (multiline — scan full text)
    # Pattern: ](optional-params) -> optional-return { ... }()
    iile_pattern = re.compile(
        r"\]\s*\([^)]*\)\s*(?:->[^{]+)?\s*\{[^}]*\}\s*\(\)",
        re.DOTALL
    )
    for match in iile_pattern.finditer(full_text):
        # Find the line number
        line_num = full_text[:match.start()].count("\n") + 1
        context = full_text[match.start():match.end()].split("\n")[0].strip()[:80]
        issues.append(
            f"  LINT [{file_path}:{line_num}] IILE: "
            f"immediately invoked lambda — extract to a named variable or function. "
            f"Line: {context}"
        )

    return issues


def main():
    raw = sys.stdin.read()
    if not raw.strip():
        sys.exit(0)

    try:
        data = json.loads(raw)
    except json.JSONDecodeError:
        sys.exit(0)

    tool_input = data.get("tool_input", {})
    file_path = tool_input.get("file_path", "")

    # Only lint C++ files
    if not file_path.endswith((".h", ".cpp", ".inl")):
        sys.exit(0)

    # Get the content being written
    content = tool_input.get("new_string") or tool_input.get("content") or ""
    if not content:
        sys.exit(0)

    lines = content.split("\n")
    issues = check_lines(lines, file_path)

    if issues:
        print(f"C++ lint ({len(issues)} issue{'s' if len(issues) != 1 else ''}):")
        for issue in issues:
            print(issue)
        print()
        print("Fix these before proceeding. These patterns are always wrong in this codebase.")

    sys.exit(0)


if __name__ == "__main__":
    main()

File: container/hooks/lint-cpp-diff.mjs
#!/usr/bin/env node
/**
 * PreToolUse lint hook for Edit/Write on C++ files.
 *
 * Checks the new content (from Edit's new_string or Write's content) against
 * mechanical rules that are always wrong in Unreal Engine C++. Returns feedback
 * lines that the agent sees immediately.
 *
 * Input: JSON on stdin with tool_input containing either:
 *   - new_string (Edit tool)
 *   - content (Write tool)
 *   - file_path
 *
 * Exit 0: no issues (tool proceeds)
 * Exit 2 + stdout: issues found (tool is BLOCKED, agent sees feedback and must fix)
 */

import { fileURLToPath } from 'node:url';

/**
 * Check lines of C++ code for lint issues.
 * @param {string[]} lines
 * @param {string} filePath
 * @returns {string[]}
 */
export function checkLines(lines, filePath) {
  const issues = [];

  // Track multiline state for IILE detection
  const fullText = lines.join('\n');

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const lineNum = i + 1;
    const stripped = line.trim();

    // Skip comments and preprocessor
    if (stripped.startsWith('//') || stripped.startsWith('#') || stripped.startsWith('/*') || stripped.startsWith('*')) {
      continue;
    }

    // Rule 1: East-const violation
    // Match "const Type&" or "const Type*" but not "const_cast" or "constexpr"
    if (/\bconst\s+(?!cast|expr|eval)\w+[\s*&]/.test(line)) {
      const match = line.match(/\bconst\s+(\w+)\s*[&*]/);
      if (match) {
        const typename = match[1];
        if (!['cast', 'expr', 'eval', 'override', 'noexcept'].includes(typename)) {
          issues.push(
            `  LINT [${filePath}:${lineNum}] East-const: ` +
            `'const ${typename}' should be '${typename} const'. ` +
            `Line: ${stripped.slice(0, 80)}`
          );
        }
      }
    }

    // Rule 2: const_cast is banned
    if (line.includes('const_cast')) {
      issues.push(
        `  LINT [${filePath}:${lineNum}] const_cast: ` +
        `const_cast is banned — fix the const-correctness of the API instead. ` +
        `Line: ${stripped.slice(0, 80)}`
      );
    }

    // Rule 3: Anonymous namespaces (break unity builds)
    if (/\bnamespace\s*\{/.test(line)) {
      issues.push(
        `  LINT [${filePath}:${lineNum}] Anonymous namespace: ` +
        `breaks unity builds. Before adding a named namespace like Resort::X::Private, ` +
        `check: (1) does this helper already exist elsewhere in the module? ` +
        `(2) should it be exposed in the header in a proper namespace instead? ` +
        `Line: ${stripped.slice(0, 80)}`
      );
    }

    // Rule 4: Greedy lambda captures [&] or [=]
    // Two branches match Python's structure: one for (params) lambdas, one for {body} no-arg lambdas
    if (/\[&\]\s*\(/.test(line) || /\[=\]\s*\(/.test(line)) {
      issues.push(
        `  LINT [${filePath}:${lineNum}] Greedy capture: ` +
        `use explicit captures instead of [&] or [=]. ` +
        `Line: ${stripped.slice(0, 80)}`
      );
    } else if (/\[&\]\s*\{/.test(line) || /\[=\]\s*\{/.test(line)) {
      issues.push(
        `  LINT [${filePath}:${lineNum}] Greedy capture: ` +
        `use explicit captures instead of [&] or [=]. ` +
        `Line: ${stripped.slice(0, 80)}`
      );
    }

    // Rule 5: Raw new (outside blessed functions)
    // Strip string literals before checking for raw new
    const lineNoStrings = line.replace(/"[^"]*"/g, '').replace(/TEXT\s*\([^)]*\)/g, '');
    const newMatch = lineNoStrings.match(/\bnew\s+([A-Z]\w+)/);
    if (newMatch) {
      const blessed = ['NewObject', 'MakeShared', 'MakeUnique', 'MakeShareable',
        'CreateDefaultSubobject', 'placement', 'operator'];
      if (!blessed.some(b => lineNoStrings.includes(b))) {
        issues.push(
          `  LINT [${filePath}:${lineNum}] Raw new: ` +
          `use NewObject<T>, MakeShared<T>, or MakeUnique<T> instead. ` +
          `Line: ${stripped.slice(0, 80)}`
        );
      }
    }

    // Rule 6: Multiple declarations on one line
    if (/^\s*\w[\w:<>*&\s]+\s+\w+\s*,\s*\w+\s*[;=]/.test(line)) {
      const beforeComma = line.split(',')[0];
      if (!beforeComma.includes('(') && !stripped.startsWith('for') && !stripped.startsWith('template')) {
        issues.push(
          `  LINT [${filePath}:${lineNum}] Multiple declarations: ` +
          `declare one symbol per line. ` +
          `Line: ${stripped.slice(0, 80)}`
        );
      }
    }

    // Rule 7: Uninitialised TSharedRef member field
    if (/\bTSharedRef\s*<[^>]+>\s+\w+\s*;/.test(line)) {
      if (!line.includes('=') && !line.includes('{')) {
        issues.push(
          `  LINT [${filePath}:${lineNum}] Uninitialised TSharedRef: ` +
          `TSharedRef has no null state — initialise with MakeShared<T>() ` +
          `or use TSharedPtr if initialisation is deferred. ` +
          `Line: ${stripped.slice(0, 80)}`
        );
      }
    }
  }

  // Rule 8: IILE detection (multiline — scan full text)
  const iilePattern = /\]\s*\([^)]*\)\s*(?:->[^{]+)?\s*\{[^}]*\}\s*\(\)/gs;
  let iileMatch;
  while ((iileMatch = iilePattern.exec(fullText)) !== null) {
    const lineNum = fullText.slice(0, iileMatch.index).split('\n').length;
    const context = fullText.slice(iileMatch.index, iileMatch.index + iileMatch[0].length).split('\n')[0].trim().slice(0, 80);
    issues.push(
      `  LINT [${filePath}:${lineNum}] IILE: ` +
      `immediately invoked lambda — extract to a named variable or function. ` +
      `Line: ${context}`
    );
  }

  return issues;
}

function main() {
  let raw = '';
  process.stdin.setEncoding('utf8');
  process.stdin.on('data', (chunk) => { raw += chunk; });
  process.stdin.on('end', () => {
    if (!raw.trim()) {
      process.exit(0);
    }

    let data;
    try {
      data = JSON.parse(raw);
    } catch {
      process.exit(0);
    }

    const toolInput = data.tool_input || {};
    const filePath = toolInput.file_path || '';

    // Only lint C++ files
    if (!filePath.endsWith('.h') && !filePath.endsWith('.cpp') && !filePath.endsWith('.inl')) {
      process.exit(0);
    }

    // Get the content being written
    const content = toolInput.new_string || toolInput.content || '';
    if (!content) {
      process.exit(0);
    }

    const lines = content.split('\n');
    const issues = checkLines(lines, filePath);

    if (issues.length > 0) {
      const s = issues.length !== 1 ? 's' : '';
      console.log(`C++ lint (${issues.length} issue${s}):`);
      for (const issue of issues) {
        console.log(issue);
      }
      console.log('');
      console.log('Fix these before proceeding. These patterns are always wrong in this codebase.');
      process.exit(2);
    }

    process.exit(0);
  });
}

// Run main when executed directly (not imported)
const isMain = process.argv[1] === fileURLToPath(import.meta.url);
if (isMain) {
  main();
}
